### ISSUE TITLE
calibrate a new model

### ISSUE THREAD

**Issue by jlaroche** [nanoporetech/megalodon#184]
Hi,
I created a new model using taiyaiki. I used the new dumped model with megalodon and I got the following error message:

ERROR: No default modified base calibration file found for guppy config: dna_r9.4.1_450bps_flipflop.cfg

Then I inserted the option "--disable-mod-calibration" in my megalodon command and everything works very well. So my model is correct.

Unfortunately, I don't understand the way to process to the calibration with "megalodon_extras calibrate" command. I look for example commands but did not find anyone. I think I have to use "megalodon_extras calibrate generate_modified_base_stats" and "megalodon_extras calibrate modified_bases" but I don't understand what data to pass to the software.

Thanks in advance

Jerome

**Comment by marcus1487** [comment]
See modified base calibration [docs here](https://nanoporetech.github.io/megalodon/modbase_training.html#megalodon-calibration) and more detailed description of the calibration command inputs in the [docs here](https://nanoporetech.github.io/megalodon/extras_calibrate.html#megalodon-extras-calibrate-generate-modified-base-stats). If there is any confusion after reading these docs please post specific questions in this thread.

**Comment by jlaroche** [comment]
Hi Marcus,

I have already read this doc. 
The error message is for the guppy config file "dna_r9.4.1_450bps_flipflop.cfg"
I don't see in the "megalodon_extras calibrate" options, how to pass my new model and the guppy config file.

What is the content of "--control-megalodon-results-dir" ?

**Comment by marcus1487** [comment]
The calibration step takes the theoretical neural network modified base probabilities and calibrates these against ground truth modified base observations from the provided samples to the calibration commands. As noted in the calibration docs: "Using [the --control-megalodon-results-dir] option assumes that all modified base statistics in --control-megalodon-results-dir represent canonical bases and all statistics in the main Megalodon results directory represent modified bases."

In short megalodon provides two ways to pass in the ground truth. The first requires two samples, one containing completely modified results (likely at a motif of interest as all modified positions does not generalize well) and a second control sample where all observations are canonical bases. The second allows the specification of a file containing reference sites from which megalodon can extract modified and control positions from a single sample for calibration.

The new model is passed to the calibrate commands via the megalodon results directories that are passed to the command. The megalodon results provided should be run with this new model in order to calibrate the modified base results for that model.

**Comment by marcus1487** [comment]
Note that you can see docs for the structure of each argument via the command line `megalodon_extras calibrate generate_modified_base_stats -h`. This may be helpful here as well.

**Comment by jlaroche** [comment]
Thank you for your answer. Here are the data I have in hand:
a negative control (PCR), a fully methylated positive control and several other samples.

Looks like I need to get first megalodon results with my new model for both positive and negative controls (with the --disable-mod-calibration option)

After this, I should run the "megalodon_extras calibrate generate_modified_base_stats" command to get a calibration file.

Am I correct if I set the parameters like this:

megalodon_extras calibrate generate_modified_base_stats
--control-megalodon-results-dir megalodon_negative_control_results
<megalodon_results_dir> megalodon_pos_control_results

Following with:
megalodon_extras calibrate modified_bases

To finally run megalodon again (without the --disable-mod-calibration option).

Do I understand well the process?

Thanks

**Comment by marcus1487** [comment]
Yes this is generally the correct structure for generating a new calibration file. One last step is to pass the calibration file to megalodon via the `--mod-calibration-filename` argument.

I would also note that a fully modified sample is generally not a good training or calibration sample. Alternatively a sample with some modified bases in known contexts is a better starting sample (though still needs more representative training data). For example for the CpG 5mC models we use the M.SssI methylase to methylate all Cs in CG contexts as a starting point. A sample using PCR with canonical Cs swapped out for 5mC would not work nearly as well at detecting modifications in a sample with strands containing a fraction of modified bases (the most likely use case for native samples). This is because when all bases are swapped out for a modified base essentially every level is shifted through the signal. This is much easier for the neural networks to detect, but is also very different from the problem of detecting some modified base within a stretch of canonical bases. The production of modified bases training datasets which are able to accurately call native modified bases is quite a bit more complicated. We are working distill this process and release the full training procedure for this, but do not have it in order to release at this point.

I hope this helps, but please do reach out with any further questions.
